### PR TITLE
ENH: Refactor actor.odf_slicer for increased performances

### PR DESCRIPTION
**Pull request by CHrlS98**
Hi,
I did some refactoring of `actor.odf_slicer` with the objective to make it faster and less memory greedy.
Basically:
* I replaced `for` loops by `numpy` array operations;
* I added the option to supply ODFs in SH coefficients, when given a B matrix;
* I added a method to change the sphere used for visualization when using SH coefficients;
* I changed the `numpy` method called in `utils.set_polydata_triangle` for one that completes faster;
* I applied the affine transform to the grid as well as to the ODFs themselves, as suggested [here](https://github.com/fury-gl/fury/issues/10);
* I moved the class `OdfSlicerActor` to the new module `fury.actors.odf_slicer` to make some space in `fury.actor` module;
* I wrote tests.

If there is some ODF data available somewhere, I would gladly add an example too.

Below are some screenshots showing new functionalities.

**Real-time ODF volume slicing**
real-time-slicing

**Changing sphere resolution during execution**
update-sphere

### CODE REVIEW

**Review by skoudoro — COMMENTED** [review]
Thank you for this nice update @CHrlS98. 

We will look into this. For backward compatibility, we might need to keep the same order of all function parameters. I wonder why the order has been changed. We do not want to break some scripts so we will need to find a way to warn the users. 

Concerning the move to a new module, we plan to do it for all actors as you can see in this issue: https://github.com/fury-gl/fury/issues/64.

Concerning example, there is no ODF data available somewhere. Feel free to generate it, and create a PR in https://github.com/fury-gl/fury-data.  All our example datas are in this place. Then, we create a specific fetcher. 

What is the size of your volume in your demo gif?

Thank again for this good work

**Comment by CHrlS98** [comment]
Hi @skoudoro 
The primary reason why I changed the order of arguments is because I feel like `sphere` should be mandatory, as it crashes when it is not supplied.

 I will look into `fury-data` and add an ODF volume for examples.

The volume shown in the screenshots is a 111 x 105 x 138 order 8 fODF SH coefficients volume of 133.0 MB.

Thanks

**Comment by skoudoro** [comment]
> The primary reason why I changed the order of arguments is because I feel like sphere should be mandatory, as it crashes when it is not supplied.

I agree that the sphere should be defined. So maybe we can just provide a default value instead of `None` and keep the parameters in order

**Comment by CHrlS98** [comment]
> I agree that the sphere should be defined. So maybe we can just provide a default value instead of `None` and keep the parameters in order

The thing is the method expects a `DIPY Sphere`, but `DIPY` is not a dependency of `FURY`. Therefore we can't set the default to some `DIPY` sphere. We would need some workaround to supply a sphere without requiring `DIPY`.

I'll try to think of something.

**Comment by skoudoro** [comment]
From `DIPY SPHERE`,  you need only the vertices and faces. However, this is already present in FURY. 

look at `[fury.primitive.prim_sphere](https://github.com/fury-gl/fury/blob/master/fury/primitive.py#L256)`, it will help.

**Comment by CHrlS98** [comment]
Nice, did not know that, adding a default will be easier than I thought.
I'll do it soon enough!

**Comment by CHrlS98** [comment]
Hi,
I switched the parameter order back to what it was and added a default sphere. I also made a PR [here](https://github.com/fury-gl/fury-data/pull/7) to make available an ODF volume I'll use for my example.

File: fury/actors/odf_slicer.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+                rescaled = sf - sf.min(axis=-1, keepdims=True)
+                rescaled[range_sf > 0] /= range_sf[range_sf > 0][..., None]
+                all_colors =\
+                    create_colormap(rescaled.ravel(), self.colormap) * 255
+            else:
+                all_colors = np.tile(np.array(self.colormap).reshape(1, 3),
+                                     (sf.shape[0]*sf.shape[1], 1))
+        else:
+            all_colors = np.tile(np.abs(self.vertices)*255, (len(sf), 1))
+        return all_colors.astype(np.uint8)
```
Comment (skoudoro) [inline fury/actors/odf_slicer.py:257]: I wonder if this block can be an independent function. I feel like I saw something similar in our codebase. I just can not remember where....

File: fury/actors/odf_slicer.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        sph_dirs = self._get_sphere_directions()
+        sf = self._get_sf(self.mask)
+
+        all_vertices = self._get_all_vertices(offsets, sph_dirs, sf)
+        all_faces = self._get_all_faces(len(offsets), len(sph_dirs))
+        all_colors = self._generate_color_for_vertices(sf)
+
+        set_polydata_triangles(polydata, all_faces)
+        set_polydata_vertices(polydata, all_vertices)
+        set_polydata_colors(polydata, all_colors)
```
Comment (skoudoro) [inline fury/actors/odf_slicer.py:179]: there is a lot of deep copy here. I think this part can be easily optimized. There are some examples in the `network` examples.

For now, it is fine, but can you add a small note/TODO. It deserves its own PR

**Review by skoudoro — COMMENTED** [review]
Hi @CHrlS98,

The code is really clean, Thank you for that. The tutorial is working well. I just have a small comment below but apart from that, I think it is ready to go.

After merging, I will tweet your gif to promote your nice work, I think it can be useful for many people and bring other contributors.

Thanks! 

ps: We will fix the design of the combobox.

File: fury/actors/odf_slicer.py (hunk 1 above)
Comment (CHrlS98) [reply fury/actors/odf_slicer.py:257]: @skoudoro How about we move it to `colormap.py` so it is some colormap utility function for generating a color array given some values and a colormap?

**Comment by CHrlS98** [comment]
@skoudoro Thanks for the nice comment,
I addressed your requests but didn't do anything about `_generate_color_for_vertices` yet. I think it could make sense to move it to `colormap.py` although it isn't clear to me what other methods could benefit from this.

On another note, I did a new GIF that I think you'll prefer. I removed the combo box and added transparency to the sagittal and axial slices. I also disabled the mouse capture.
better-slicing

**Comment by skoudoro** [comment]
Ok, I tested it and it is good to go. 

Concerning `_generate_color_for_vertices`, I believe it was `line` or `surface` actors. I will create an issue for that. We have to review color and colormap in general. Sometimes the initialization is between [0-1] and sometimes [0-255]. There is a uniformization task to do.

Thank you for the updated gif. What is your Twitter handle?

Concerning checkboxes, you should look at dipy_horizon. we transform the `X, Y, Z` labels to flat buttons which avoid having checkboxes. The buttons hide/show the slices and sliders.

**Comment by CHrlS98** [comment]
@skoudoro Alrighty thanks,
I'll look into `dipy_horizon` for flat buttons :+1: 
My twitter: **@CHrlS_98**

See you around!
